### PR TITLE
Add a full screen modal component

### PR DESCRIPTION
**Pull request by youknowriad**
This PR restores an experiment I did a long time ago to allow full screen modals.
The main use-case for this is going to be the "template switch" modal explored in #28390 
For the current PR, I applied the styles to the "Preferences" modal as well. I personally like it but it folks think it's not a good fit, we can restore the style before merging.

**Testing instructions**

 - Check different modals: Preferences, Keyboard shortcuts 
 - Make sure they look great in both mobile and web

### CODE REVIEW

**Comment by jasmussen** [comment]
I understand how this component might benefit work in #28390 — it seems a legitimate use case.

But although the following is pretty, I don't think it's actually an improvement over what we have:

Screenshot 2021-01-29 at 10 09 28

It's cleaner, but it severs the context you have by seeing the editor around it. It may be fine? But I'd love some more sanity checks, and happy to defer.

But one path forward is to build this component for use with, as mentioned, 28390, and use it there, but not for preferences or keyboard shortcuts. Or does it replace entirely the current modal system?

**Comment by paaljoachim** [comment]
It is important to keep the context where one is when one clicks example the preferences/options, templates, etc etc. This makes full screen modals difficult because they shatter any contextual relation to the page/post the user is on.
For instance. I made issues using modals for editing/viewing 
Reusable blocks: https://github.com/WordPress/gutenberg/issues/26408#issuecomment-724711296
and
patterns: https://github.com/WordPress/gutenberg/issues/26905#issue-741047109
and
Revisions panel: https://github.com/WordPress/gutenberg/issues/25270

We could for instance add a similar method for templates. 

The context of a modal still keeps the user related to where they are. Modals can for instance take over most of the screen still keeping the outside area intact and semi greyed/blacked out where one can see the Gutenberg area and still keep a connection to where one came from.

**Comment by jameskoster** [comment]
I quite like how iOS handles modals, it does a good job of maintaining context without sacrificing too much space or creating a cramped sensation. I made a [prototype](https://cdpn.io/jameskoster/debug/abdpGXb#) ages ago to see how it translated to desktop. The details probably need fine tuning, but I think it works quite well. Obviously scales down to mobile nicely as well :D

**Review by ntsekouras — COMMENTED** [review]
I really like this! Thanks Riad!

>It's cleaner, but it severs the context you have by seeing the editor around it.

I echo Joen for **not** making the `Preferences` full screen, but `Keyboard shortcuts` seems much better fullscreen IMO.

Either way this can be merged without changing any existing component yet - though I really liked fullscreen Keyboard shortcuts :)

The only thing that seems just a bit off to me is the animation. I'm not sure if I'm just accustomed to the current one that is different.. 🤔

**Comment by paaljoachim** [comment]
Thanks for working on this Riad!
It is nice to be able to have a full screen modal component in the Block editor. As there are likely use cases where it becomes the natural choice to use.

I just want to add another modal that we do use a lot.
When we open the Media Library through a block to add media.
Screen Shot 2021-02-05 at 14 32 30
It is almost full screen but it still retains some context to the Gutenberg screen.

To me it seems we need to define the various types of modal components that we have. So we know which of these would be natural to use for various features that need a modal. Maybe even a doc showing the various modal types.

**Comment by youknowriad** [comment]
Thanks for your thoughts everyone, I removed  the full screen modal design from the "preferences" modal. We can focus on the component for now in order to land it and use it in #28390 instead

**Comment by jameskoster** [comment]
I was getting errors when building this PR so played around in the Inspector to create:

full-modal

As mentioned above, the modal being full width has benefits on smaller screens where horizontal space is at a premium:



The animation and void at the top help maintain the relationship between screen and modal.

**Comment by paaljoachim** [comment]
It looks really interesting @jameskoster  
Btw here are some initial sketches by Tammie:
https://github.com/WordPress/gutenberg/issues/19128#issue-537589282

**Comment by jameskoster** [comment]
Oh, yes, ignore the contents of the modal there. I agree that _preferences_ should remain in a "regular" modal.

**Comment by youknowriad** [comment]
@jameskoster Happy to try to implement that if that's the direction we want to go and if it fits #28390

Can you share the custom CSS you had to write to make that work (even hacky one)

**Comment by paaljoachim** [comment]
It would be great if we can decide/define standard modals to use. Looking at the various types of modals in use today. So that we can create a consistency of visual items that popup on the screen.

Such as for the options/preferences modal all the way up in size to the Add Media modal.
Perhaps the Add Media Modal will fit with the template issue mentioned?

**Comment by youknowriad** [comment]
I think we shouldn't be using "Add Media" kind of modals at all, that's backbone modals which doesn't fit great the components library of Gutenberg. Ideally, we should do the opposite, at some point in the future, deprecate the media modal as it exists today and rely on `@wordpress/components`

**Comment by jameskoster** [comment]
Ah I'm afraid I closed the tab. For the "background" it was basically a scale transform on `.modal-open #wpwrap` (we'd probably need a `full-screen-modal-open` class so that normal modals are excluded from the effect).

Then the modal itself had a `top: 100%;` position when hidden, and `top: X;` when visible.

Happy to dig in to the css if we can fix the branch up. Probably just needs a rebase.

**Comment by youknowriad** [comment]
>  For the "background" it was basically a scale transform on .modal-open #wpwrap

This is the issue here, the problem is that we don't really know that `.modal-open #wpwrap` exists, it's a wp specific div

**Comment by youknowriad** [comment]
Ok found a hack, let me know how it looks (try keyboard shortcuts) and feel free to tweak the styles and animations.

**Comment by jameskoster** [comment]
Nice! I think there are a few visual optimisations we can make. I'll look into this on Monday 👍

**Comment by jameskoster** [comment]
Made some style and animation adjustments:

https://user-images.githubusercontent.com/846565/107952042-206eb780-6f91-11eb-8855-689fe8797862.mp4

It's a shame we can't have the animation reverse when the modal is closed. It is a little jarring to jump immediately back with no transition.

File: packages/components/src/modal/style.scss [hunk 1]
```diff
@@ -134,17 +133,46 @@
...
-	transition: transform 0.5s;
-	transform: scale(0.9);
+body.modal-open.is-fullscreen-modal {
+	background: $gray-900;
+
+	> *:not(.components-modal__portal) {
+		animation: components-modal__content-scale-back 0.4s cubic-bezier(0.165, 0.84, 0.44, 1) forwards;
+	}
+
+	.interface-interface-skeleton__editor {
```
Comment (youknowriad) [inline packages/components/src/modal/style.scss:143]: This is not something we can target here, we don't know what's rendered on the page on a generic component like "Modal"

Comment (jameskoster) [reply packages/components/src/modal/style.scss:143]: I'll remove 👍

**Review by jameskoster — APPROVED** [review]

**Comment by youknowriad** [comment]
Closing as we might not need this anymore in the latest template switching designs.

**Pull request closed without merging**